packages/multi-test/src/staking.rs

```diff
@@ -15,12 +29,240 @@ pub enum StakingSudo {
+        &self,
+        storage: &mut dyn Storage,
+        account: &Addr,
+        amount: Vec<Coin>,
+    ) -> AnyResult<()> {
+        let mut storage = prefixed(storage, NAMESPACE_STAKING);
+        self.set_balance(&mut storage, account, amount)
+    }
+
+    fn get_stakes(&self, storage: &dyn Storage, account: &Addr) -> AnyResult<Vec<Coin>> {
+        let val = STAKES.may_load(storage, account)?;
+        Ok(val.unwrap_or_default().into_vec())
+    }
+
+    fn set_balance(
+        &self,
+        storage: &mut dyn Storage,
+        account: &Addr,
+        amount: Vec<Coin>,
+    ) -> AnyResult<()> {
+        let mut stake = NativeBalance(amount);
+        stake.normalize();
+        STAKES.save(storage, account, &stake).map_err(Into::into)
+    }
+
+    fn add_stake(
+        &self,
+        storage: &mut dyn Storage,
+        to_address: Addr,
+        amount: Vec<Coin>,
+    ) -> AnyResult<()> {
+        let amount = self.normalize_amount(amount)?;
+        let b = self.get_stakes(storage, &to_address)?;
+        let b = NativeBalance(b) + NativeBalance(amount);
+        self.set_balance(storage, &to_address, b.into_vec())
+    }
+
+    fn remove_stake(
+        &self,
+        storage: &mut dyn Storage,
+        from_address: Addr,
+        amount: Vec<Coin>,
+    ) -> AnyResult<()> {
+        let amount = self.normalize_amount(amount)?;
+        let a = self.get_stakes(storage, &from_address)?;
+        let a = (NativeBalance(a) - amount)?;
+        self.set_balance(storage, &from_address, a.into_vec())
+    }
+
+    /// Filters out all 0 value coins and returns an error if the resulting Vec is empty
+    fn normalize_amount(&self, amount: Vec<Coin>) -> AnyResult<Vec<Coin>> {
+        let res: Vec<_> = amount.into_iter().filter(|x| !x.amount.is_zero()).collect();
+        if res.is_empty() {
+            bail!("Cannot transfer empty coins amount")
+        } else {
+            Ok(res)
+        }
+    }
+}
+
+impl Staking for StakeKeeper {}
+
+impl Module for StakeKeeper {
+    type ExecT = StakingMsg;
+    type QueryT = StakingQuery;
+    type SudoT = StakingSudo;
 
-pub type FailingDistribution = FailingModule<DistributionMsg, Empty, Empty>;
+    fn execute<ExecC, QueryC>(
+        &self,
+        _api: &dyn Api,
+        storage: &mut dyn Storage,
+        _router: &dyn CosmosRouter<ExecC = ExecC, QueryC = QueryC>,
+        _block: &BlockInfo,
+        sender: Addr,
+        msg: StakingMsg,
+    ) -> AnyResult<AppResponse> {
+        let mut staking_storage = prefixed(storage, NAMESPACE_STAKING);
+        match msg {
+            StakingMsg::Delegate { validator, amount } => {
+                let events = vec![Event::new("delegate")
+                    .add_attribute("recipient", &validator)
+                    .add_attribute("sender", &sender)
+                    .add_attribute("amount", format!("{}{}", amount.amount, amount.denom))];
+                self.add_stake(&mut staking_storage, sender, vec![amount])?;
```

Comment:
TODO: we need to deduct from the account that was validating (from the bank module).
I am pretty sure this is possible, I just need some time to do this (and then I will push it)